app/services/offender_service.rb

```diff
@@ -1,27 +1,80 @@
 # frozen_string_literal: true
 
 class OffenderService
-  def self.get_offender(offender_no)
-    HmppsApi::PrisonApi::OffenderApi.get_offender(offender_no).tap { |o|
-      next false if o.nil?
+  class << self
+    def get_offender(offender_no)
+      # use find_or_create_by here for performance, but might still have be a small race condition
+      # This isn't find_or_create_by! because our offender_no might be invalid - we want to return nil rather than crash
+      offender = Offender.find_or_create_by(nomis_offender_id: offender_no)
+      if offender
+        api_offender = HmppsApi::PrisonApi::OffenderApi.get_offender(offender_no)
+        if api_offender
+          api_offender.load_main_offence
+          prison = Prison.find_by(code: api_offender.prison_id)
+          # ignore offenders in prisons that we don't know about
+          MpcOffender.new prison: prison, offender: offender, prison_record: api_offender if prison
+        end
+      end
+    end
 
-      # use find_or_create_by! here for performance, but might still have be a small race condition
-      offender = Offender.find_or_create_by!(nomis_offender_id: o.offender_no)
-      o.load_case_information(offender.case_information)
+    def get_offenders_for_prison(prison)
+      OffenderEnumerator.new(prison)
+    end
 
-      o.load_main_offence
-    }
-  end
+    class OffenderEnumerator
+      include Enumerable
+      FETCH_SIZE = 200 # How many records to fetch from nomis at a time
+
+      def initialize(prison)
+        @prison = prison
+      end
+
+      def each
+        first_page = HmppsApi::PrisonApi::OffenderApi.list(@prison.code, 0, page_size: FETCH_SIZE)
+        offenders = first_page.data
+        enrich_offenders(offenders).each { |offender| yield offender }
+
+        1.upto(first_page.total_pages - 1).each do |page_number|
+          offenders = HmppsApi::PrisonApi::OffenderApi.list(
+            @prison.code,
+            page_number,
+            page_size: FETCH_SIZE
+          ).data
+
+          enrich_offenders(offenders).each { |offender| yield offender }
+        end
+      end
+
+      def enrich_offenders(offender_list)
+        nomis_ids = offender_list.map(&:offender_no)
+        offenders = Offender.
+          includes(case_information: [:responsibility, :early_allocations, :local_delivery_unit]).
+          where(nomis_offender_id: nomis_ids)
```

Comment:
### Performance optimisation (todo?)

Presumably we'll also want to be eager loading things like `ParoleRecord` (for PRDs) once that PR is in...

For example, PRD is likely to be needed (I think?) when calculating handover dates/responsibility. And that happens when filtering offenders (e.g. approaching handover).